interfaces/sourcegen/sourcegen/csharp/_SourceGenerator.py

```diff
@@ -0,0 +1,309 @@
+            # from which we determine the appropriate C# type
+            prop_type = self._prop_type_map[getter.params[-1].p_type]
+
+        setter = self._known_funcs.get(clazz + "_set" + c_name.capitalize())
+
+        if prop_type in ['int', 'double']:
+            text = f'''
+                public {prop_type} {cs_name}
+                {{
+                    get => InteropUtil.CheckReturn(
+                        LibCantera.{getter.name}(_handle));'''
+
+            if(setter):
+                text += f'''
+                    set => InteropUtil.CheckReturn(
+                        LibCantera.{setter.name}(_handle, value));'''
+
+            text += '''
+                }
+            '''
+        elif (prop_type == 'string'):
+            p_type = getter.params[1].p_type
+
+            # for get-string type functions we need to look up the type of the second
+            # (index 1) param for a cast because sometimes it's an int and other times
+            # its a nuint (size_t)
+            text = f'''
+                public unsafe string {cs_name}
+                {{
+                    get => InteropUtil.GetString(40, (length, buffer) =>
+                        LibCantera.{getter.name}(_handle, ({p_type}) length, buffer));
+            '''
+
+            if(setter):
+                text += f'''
+                    set => InteropUtil.CheckReturn(
+                        LibCantera.{setter.name}(_handle, value));'''
+
+            text += '''
+                }
+            '''
+        else:
+            raise ValueError(f'Unable to scaffold properties of type {prop_type}!')
+
+        return(normalize_indent(text))
+
+
+    def generate_source(self, incl_file: os.DirEntry, funcs: list[Func]):
+        cs_funcs = [self._convert_func(f) for f in funcs]
+
+        functions_text = '\n\n'.join((self._get_function_text(f) for f in cs_funcs))
+
+        interop_text = normalize_indent(f'''
+            {normalize_indent(self._preamble)}
+
+            using System.Runtime.InteropServices;
+
+            namespace Cantera.Interop;
+
+            static partial class LibCantera
+            {{
+                {normalize_indent(functions_text)}
+            }}
+        ''')
+
+        with open(self._out_dir + 'Interop.LibCantera.'
+                + incl_file.name + '.g.cs', 'w') as f:
+            f.write(interop_text)
+
+        handles = [(name, del_clazz) for _, name, _, del_clazz in cs_funcs if del_clazz]
+
+        if not handles:
+            return
+
+        handles_text = '\n\n'.join((self._get_base_handle_text(h) for h in handles))
+
+        handles_text = normalize_indent(f'''
+            {normalize_indent(self._preamble)}
+
+            namespace Cantera.Interop;
+
+            {normalize_indent(handles_text)}
+        ''')
+
+        with open(self._out_dir + 'Interop.Handles.'
+                + incl_file.name + '.g.cs', 'w') as f:
+            f.write(handles_text)
+
+
+    def finalize(self):
```

Comment:
What's the difference between what this function is doing and what `generate_source()` is doing? This looks like its still generating source files.

Reply:
Indeed. In this initial pass at the API, `generate_source()` writes files that can be scaffolded from one CLib header file. It also builds up some lists for files that need info from multiple headers, such that they are written later, during `finalize`. My thought was that it was good to write files as soon as possible, but if memory and such aren't really a concern, one could also just parse _all_ the headers and call `generate_source()` a single time, obviating the need for a (public) `finalize()` method.

Reply:
I'd say that optimizing for memory usage and such at this point is a little premature. I'm more interested in getting the API here correct, not least because this is developer-facing only and will only run once for each interface ☺️

I think there's a utility for a `finalize()` function, but I just want to be clear and consistent about what's happening in each stage.

Reply:
Ok so maybe the “orchestrator” script just parses all the files at once and calls the sub-packages’ implementation of `generate_source` one time as in “Okay, here are all your symbols for everything”. But in that case, a secondary public `finalize` method seems redundant (it feels a little like I wrote a C-style program with a `main` and an `other_main` function). In this case, we could make it explicit that there is a single entry-point/handover into the lang-specific generator with a different name. The language-specific implemenation would of course not just be one giant function but have whatever methods are appropriate for that language.

Reply:
> Ok so maybe the “orchestrator” script just parses all the files at once and calls the sub-packages’ implementation of `generate_source` one time as in “Okay, here are all your symbols for everything”.

Yes, I think that makes sense and simplifies things.

> But in that case, a secondary public `finalize` method seems redundant

That's probably true... I think there's still a possibility that some languages will want to be able to do something after writing all the files out (say, a compiling step or a checking step). That said, adding methods to the API later will be easier than removing them, so I think it makes sense to leave out `finalize()` for now and add it later if we feel a need for it.